chore: new commit message policy using conventional-changelog

vlapo opened the pull request:
Hi folks. 
Here is my simple proposal how to switch from annoying "update changelog.md on every PR" into generate changelog only once on release new version using `conventional-changelog`. However this method will require little bit more attention from reviewers, moderators and person who will merge PRs and release new version (in our case @pleerock so please if you have any comments)

## Commit messages
This PR update `CONTRIBUTING.md` and inform contributors about format of commit messages. I used texts from angular/material project and add some simple examples. I hope for start we will use at least commit messages with `type` (`scope`) and `subject/short description` of change like `fix(mysql): change mysql driver behaviour during connect....`

## PR titles
I think we should update all PR titles to format `<type>(<scope>): <short description>`. 
 - it will help at least find duplicate PRs if we use descriptive titles
 - it will be easier merge PRs as described bellow

## Merging PRs
We should use "Squash and merge" option here on GitHub. It will be use PR title as commit message and squash all commits in PR into one in our git history. We will have more clear git history.
![image](https://user-images.githubusercontent.com/2484063/57734828-2b513580-76a3-11e9-9595-167b5bed1468.png)
![image](https://user-images.githubusercontent.com/2484063/57734877-47ed6d80-76a3-11e9-84a7-78b402eeac35.png)


## Generate changelog
I am not sure if setup in this PR is one hundred percent ok, so please @pleerock comment this as much as possible. Steps for release:
 - generate dist, update version
 - run `npm run changelog`
 - previous command will generate changelog for unreleased changes into `CHANGELOG.md`. Something like:
```
# Unreleased (2019-05-14)
  

### Bug Fixes

* fix bla bla issue ([#6](https://github.com/typeorm/typeorm/ssues/6)) ([d30c780](https://github.com/vlapo/github-test/commit/d30c780)), closes [#1](https://github.com/typeorm/typeormissues/1)
```
 - it is necessary update release title with version and move new generated items into right place in `CHANGELOG.md` like:
```
## 0.2.22  (2019-05-14)


### Bug Fixes

* fix bla bla issue ([#6](https://github.com/typeorm/typeorm/ssues/6)) ([d30c780](https://github.com/vlapo/github-test/commit/d30c780)), closes [#1](https://github.com/typeorm/typeormissues/1)
```
 - commit changes

## Generate changelog 2

Another way how to generate changelog is: 
 - generate dist, update version, create git tag
 - commit/push changes and tag
 - change command to generate changelog for last release: `conventional-changelog -p angular -i CHANGELOG.md -s -r 1`
 - it will be generate new items in changelog
 - update changelog as we wish
 - commit/push new changelog e.g. with commit message `chore: update changelog for 0.x.x`

Add @Kononnable, @AlexMesser, @rustamwin

Kononnable commented on CONTRIBUTING.md:
```diff
@@ -123,6 +124,96 @@ from the main (upstream) repository:
+the SHA of the commit being reverted.
+
+### Type
+Must be one of the following:
```
I don't like changelogs which are same as formatted output of `git log`. I believe changelog should contain only info which might come handy for end users - so only feat, fix, perf types.

vlapo replied:
I agree. Angular preset I used will generate changelog only from `feat`, `fix` and `perf` as you mentioned. It will look like this changelog https://github.com/angular/components/blob/master/CHANGELOG.md.

Kononnable commented on CONTRIBUTING.md:
```diff
@@ -123,6 +124,96 @@ from the main (upstream) repository:
+            (example scopes: gulp, broccoli, npm)
+* **chore**: Other changes that don't modify `src` or `test` files
+
+### Scope
```
I think if we want to use scopes we should determine their names - otherwise there will be very common for one commit to belong to many scopes.
But If we decide to put only some commit messages into changelog scopes might become unnecessary at all.

vlapo replied:
Maybe define just `core` and all drivers as scopes for now? Or just omit this part of commit message.

Kononnable replied:
I think we could omit it for now. It will be useful when we split codebase into multiple packages(if we decide to go with monorepo approach).

Kononnable commented on CONTRIBUTING.md:
````diff
@@ -123,6 +124,96 @@ from the main (upstream) repository:
+```
+refactor(core): refactor driver API and generate separate packages per driver
+
+BREAKING CHANGE: it is necessary install new core package `npm install @typeorm/core` and specific driver `npm install @typeorm/postgres`
````
I would advice to change this example since we don't have such packages yet. It's going to happen some day, but it's not certain `@typeorm/postgres` will be an exact name.

vlapo replied:
Sure agree :) Will change this.

Kononnable commented on CONTRIBUTING.md:
```diff
@@ -123,6 +124,96 @@ from the main (upstream) repository:
 
+## <a name="commit"></a> Commit Message Guidelines
+
+We have very precise rules over how our git commit messages can be formatted.  This leads to **more
```
Should we mention here, that we want to use squash merge PR? Contributors don't really have to write proper formatted commit messages since their messages will disappear either way.

vlapo replied:
- I dont think so contributor have to know how we merge PRs.
 - also `conventional-changelog` is standard in lot of open-sources - they will learn something new
 - commit message is used as PR title so we do not have to rename PRs if contributor use good description and our format

Kononnable replied:
I don't think we will be able to do anything if contributors write not standardized commit messages. Blocking merge just because of commit messages which are discarded anyway isn't the best option.
However we can write it here - at least some folks might use standardized messages.

As for the last bullet point - this might by true, but mostly only for one commit PRs. For bigger PRs we would still have at least cpy&paste them manually. But it if saves work at least in some situations it sounds good.

vlapo replied:
Understand you. But it is necessary to write it in this document? I think we can just have some internal agreement that it is non-critical issue and we can merge also PRs with non-standard commit message format.

Kononnable commented:
In general - I like this PR. We will have to check if changelog was generated properly first time it is used(I've read somewhere that there might be some problems when generating changelog for the first time when one with different conventions already exists).

vlapo commented:
```conventional-changelog -p angular -i CHANGELOG.md -s -u```

This command will create only changelog for unreleased changes since last version tag. So it is possible to format `changelog.md` as we wish and just copy generated items into right place in changelog.md. So old changelog content will be untouched.

pleerock commented:
> I think we should update all PR titles to format <type>(<scope>): <short description>.

Is it something we should do? If yes, then do we need to do it before we merge this PR?

pleerock commented:
I'm okay to merge it, lets try it :) Just let me know what needs to be done from my side after we merge it

rustamwin commented:
> We should use "Squash and merge" option here on GitHub

:+1: Agree

Also we can add PR template.

vlapo commented:
> > I think we should update all PR titles to format (): .
> 
> Is it something we should do? If yes, then do we need to do it before we merge this PR?

I dont think so we have to do this. It was only idea.

> I'm okay to merge it, lets try it :) Just let me know what needs to be done from my side after we merge it

 - use "Squash and merge" 
 - use standard commit message format
 - run `npm run changelog` before release new version and you will see unreleased changes in changelog.md - We have to test it so I can help you with this. Just let me know on slack :)

Kononnable commented:
I think scopes are only thing blocking it from being merged. Everything else seem to be agreed on.

As for PR template - we could add it, but do we know what we want there to be? Tbh. I don't have a clue what should we put in such template. But if anyone have an idea - go ahead, we can discuss it.

vlapo commented:
Sry for delay. Updated :)

rustamwin commented:
> But if anyone have an idea - go ahead, we can discuss it.

| Q             | A
| ------------- | ---
| Is bugfix?    | yes/no
| New feature?  | yes/no
| Breaks BC?    | yes/no
| Tests pass?   | yes/no
| Fixed issues  | comma-separated list of tickets # fixed by the PR, if any
### description (optional)

Kononnable commented:
As mentioned in #4183 - should we add `commitlint` as a hook to husky to force using properly formatted commit messages? It would minify work which has to be done on merge.

@rustamwin I think first three questions will be answered by commit messages now(and most of the time PR title, because people will mostly leave default one). Tests results can be seen automatically by icons next to PR name(on the list) or just above merging section in PR page. I think people will forget to update first post after pushing changes and information might be misleading.
As for the fixed issues: I think this is a wonderful idea. We could use [special keyword syntax](https://help.github.com/en/articles/closing-issues-using-keywords) to close issues automatically after merging PR.

rustamwin commented:
> We could use special keyword syntax to close issues automatically after merging PR.

Updated.

vlapo commented:
> As mentioned in #4183 - should we add `commitlint` as a hook to husky to force using properly formatted commit messages? It would minify work which has to be done on merge.

We discussed about commit message format as optional above. Commitlint will for sure block some PRs (I think mostly first time contributors). But still I think this idea is worth to try. We can remove it anytime.

> @rustamwin I think first three questions will be answered by commit messages now(and most of the time PR title, because people will mostly leave default one). Tests results can be seen automatically by icons next to PR name(on the list) or just above merging section in PR page. I think people will forget to update first post after pushing changes and information might be misleading.

I feel it same.

> As for the fixed issues: I think this is a wonderful idea. We could use [special keyword syntax](https://help.github.com/en/articles/closing-issues-using-keywords) to close issues automatically after merging PR.

I cover this in example section:
Fix and close issue:
```
fix: resolve issues uppercase column names

Closes: #123456
```
This format of commit message will cause closing issue too.

Kononnable commented:
Looks like we're gonna forget about this PR like of some in the past which were already agreed on(test changes).

To summarize: I would like to add `commitlint` - if no one opposes. People will get some errors while trying to add commits, but we should be able to add some link to contributing guide in an error message. If they won't find it they probably will ask in slack channel or issue related to their PR.
As for PR template I will leave it as it is(at least for now). Only info about related issues seems more useful for me, but people will link it either way(and closing can be done by commit messages). However we can still discuss it(probably best in new issue/PR) - so we won't block merging this PR.

And I guess it will be the time to finally merge it, so it's not become forgotten and outdated.

pleerock commented:
From what I understand to add a new record into a CHANGELOG we just need to press "Squash and merge” and write a title that is going to be used in that record. Since its what we do (merging and editing title), why do we need to use commitlint? I think it should be our responsibility to make sure its a correct record in a changelog and we can do it easily just before merging a PR.

pleerock commented:
@vlapo lets try to merge it and see how it works for us, and improve process in the future PRs
